core/src/Cabana_LinkedCellList.hpp

```diff
@@ -410,6 +439,115 @@ void permute(
+            grid_min[0], grid_min[1], grid_min[2], grid_max[0], grid_max[1],
+            grid_max[2], dx, dx, dx );
+        cell_range = std::ceil( 1 / cell_size_ratio );
+        max_cells_dir = 2 * cell_range + 1;
+        max_cells = max_cells_dir * max_cells_dir * max_cells_dir;
+    }
+
+    //! Given a cell, get the index bounds of the cell stencil.
+    KOKKOS_INLINE_FUNCTION
+    void getCells( const int cell, int& imin, int& imax, int& jmin, int& jmax,
+                   int& kmin, int& kmax ) const
+    {
+        int i, j, k;
+        grid.ijkBinIndex( cell, i, j, k );
+
+        kmin = ( k - cell_range > 0 ) ? k - cell_range : 0;
+        kmax =
+            ( k + cell_range + 1 < grid._nz ) ? k + cell_range + 1 : grid._nz;
+
+        jmin = ( j - cell_range > 0 ) ? j - cell_range : 0;
+        jmax =
+            ( j + cell_range + 1 < grid._ny ) ? j + cell_range + 1 : grid._ny;
+
+        imin = ( i - cell_range > 0 ) ? i - cell_range : 0;
+        imax =
+            ( i + cell_range + 1 < grid._nx ) ? i + cell_range + 1 : grid._nx;
+    }
+};
+
+//---------------------------------------------------------------------------//
+//! LinkedCellList NeighborList interface.
+template <class DeviceType>
+class NeighborList<LinkedCellList<DeviceType>>
+{
+  public:
+    //! Kokkos memory space.
+    using device_type = DeviceType;
+    //! Neighbor list type.
+    using list_type = LinkedCellList<DeviceType>;
+
+    //! Get the maximum number of neighbors per particle.
+    template <class CellIndexType, class StencilType>
+    KOKKOS_INLINE_FUNCTION static std::size_t
+    maxNeighbor( const list_type& list, const CellIndexType cell,
+                 const StencilType stencil )
+    {
+        int total_count = 0;
+        for ( int p = 0; p < list.numParticles(); ++p )
+            total_count += numNeighbor( list, cell, stencil, p );
+        return total_count;
+    }
+
+    //! Get the number of neighbors for a given particle index.
+    template <class CellIndexType, class StencilType>
+    KOKKOS_INLINE_FUNCTION static std::size_t
+    numNeighbor( const list_type& list, const CellIndexType cell,
+                 const StencilType stencil, const std::size_t particle_index )
+    {
+        int total_count = 0;
+        int imin, imax, jmin, jmax, kmin, kmax;
+        stencil.getCells( cell( particle_index ), imin, imax, jmin, jmax, kmin,
+                          kmax );
+
+        // Loop over the cell stencil.
+        for ( int i = imin; i < imax; ++i )
+            for ( int j = jmin; j < jmax; ++j )
+                for ( int k = kmin; k < kmax; ++k )
+                {
+                    total_count += list.binSize( i, j, k );
+                }
+        return total_count;
+    }
+
+    //! Get the id for a neighbor for a given particle index and the index of
+    //! the neighbor relative to the particle.
+    KOKKOS_INLINE_FUNCTION
+    static std::size_t getNeighbor( const list_type& list, const std::size_t,
```

Comment:
We are leaving it up to the user to determine if this neighbor is in the radius right? The way this is structured, a particle is a neighbor if it is in the bin of a given particle or in an adjacent bin.

Reply:
Right, this requires a user to ignore based on radius, but we should consider adding new discriminators